Call into WKApplicationUtilities when handling certain types of links

bnham opened the pull request:
#### 040d74b4188067ac84d941cc83a85dab997015f4
<pre>
Call into WKApplicationUtilities when handling certain types of links
<a href="https://bugs.webkit.org/show_bug.cgi?id=265172">https://bugs.webkit.org/show_bug.cgi?id=265172</a>
<a href="https://rdar.apple.com/118667681">rdar://118667681</a>

Reviewed by Brent Fulgham and Elliott Williams.

In order to implement a particular feature, we need to call into WKApplicationUtilities when
handling certain types of navigation actions. This then calls a Swift-only API. To do this, we copy
WKApplicationUtilities.swift from WebKitAddition&apos;s include location to WebKit&apos;s DerivedSources. We
use EXCLUDED_SOURCE_FILES to exclude that file on platforms and SDKs where it doesn&apos;t exist.

This also requires us to know the top frame&apos;s origin at the point of navigation, and I couldn&apos;t find
any good way of accessing that state in decidePolicyForNavigationAction. So I modified
NavigationActionData to carry the topOrigin state.

* Source/WebKit/Configurations/WebKitSwift.xcconfig:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/NavigationActionData.h:
* Source/WebKit/Shared/NavigationActionData.serialization.in:
* Source/WebKit/UIProcess/Cocoa/NavigationState.mm:
(WebKit::NavigationState::NavigationClient::decidePolicyForNavigationAction):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::createWindow):
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
(WebKit::WebFrameLoaderClient::dispatchDecidePolicyForNavigationAction):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::didSameDocumentNavigationForFrameViaJSHistoryAPI):
(WebKit::WebLocalFrameLoaderClient::dispatchDecidePolicyForNewWindowAction):

Canonical link: <a href="https://commits.webkit.org/271405@main">https://commits.webkit.org/271405@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/34efecf62f342490f415a259593d66f68325df1d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/27910 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/6548 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/29211 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/30441 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/25472 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/8535 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/3943 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/25239 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28176 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/5313 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/23981 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/4570 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/4764 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/31129 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/25521 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/25428 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/31014 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/4769 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/2928 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/28825 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/6293 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6777 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/5215 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/5240 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham reviewed (approved):
This change looks correct to me. r=me

emw-apple commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -7543,6 +7551,7 @@
 		EB450E0E2996C7A1009724B1 /* WKWebsiteDataStoreRefPrivateMac.mm */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.cpp.objcpp; name = WKWebsiteDataStoreRefPrivateMac.mm; path = mac/WKWebsiteDataStoreRefPrivateMac.mm; sourceTree = "<group>"; };
 		EB579C3529AEBCF100894C1C /* EarlyHintsResourceLoader.cpp */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.cpp.cpp; path = EarlyHintsResourceLoader.cpp; sourceTree = "<group>"; };
 		EB579C3629AEBCF100894C1C /* EarlyHintsResourceLoader.h */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.c.h; path = EarlyHintsResourceLoader.h; sourceTree = "<group>"; };
+		EB7169D22B05D8B2005F305D /* DerivedSources/WebKit/WKApplicationUtilities.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; path = DerivedSources/WebKit/WKApplicationUtilities.swift; sourceTree = BUILT_PRODUCTS_DIR; };
```
nit: Derived sources typically go in the "Derived Sources" group, with their location set to "Relative to Group".

bnham replied:
Yeah this is definitely not the way I meant to include the file into the project. It's kind of surprising it even builds. Will fix.

emw-apple commented on Source/WebKit/DerivedSources.make:
```diff
@@ -796,3 +796,14 @@ module.private.modulemap : $(WK_MODULEMAP_PRIVATE_FILE)
+	WKApplicationUtilities.swift \
+#
+
+$(WEBKIT_ADDITIONS_SWIFT_FILES): %.swift : %.swift.in
+	cp -f $< $@
```
Not sure I understand why the input file is a `.in`, since we're just copying it. We use `.in` elsewhere to imply that there's a preprocessing / templating step. Can it just be `.swift`?

bnham replied:
I wasn't sure if a Makefile rule like `$(WEBKIT_ADDITIONS_SWIFT_FILES): %.swift : %.swift` would work, but I'll give it a shot.

bnham replied:
Yeah that just results in a circular dependency that gets dropped by make. I think I could probably make it work if I explicitly included the entire path of the source file, but I think I want to depend on `VPATH` search here so that `WKApplicationUtilities.swift` in built products gets prioritized over `WKApplicationUtilities.swift` in the SDK directory correctly.

emw-apple replied:
Seems like a reasonable trade-off!

emw-apple commented on Source/WebKit/Configurations/WebKitSwift.xcconfig:
```diff
@@ -39,7 +39,16 @@ WK_EXCLUDED_COORDINATOR_FILES_ = WKGroupSession.swift
+WK_EXCLUDED_APPLICATION_UTILITIES_FILES = $(WK_EXCLUDED_APPLICATION_UTILITIES_FILES_$(USE_INTERNAL_SDK))
+WK_EXCLUDED_APPLICATION_UTILITIES_FILES_ = WKApplicationUtilities.swift
+WK_EXCLUDED_APPLICATION_UTILITIES_FILES_YES = $(WK_EXCLUDED_APPLICATION_UTILITIES_FILES_YES_$(WK_NOT_$(WK_HAVE_APPLICATION_UTILITIES)));
+WK_EXCLUDED_APPLICATION_UTILITIES_FILES_YES_YES = WKApplicationUtilities.swift
```
I think it's easier to read this way:
```suggestion
WK_EXCLUDED_APPLICATION_UTILITIES_FILES = $(WK_EXCLUDED_APPLICATION_UTILITIES_FILES_$(WK_AND_$(USE_INTERNAL_SDK)_$(WK_HAVE_APPLICATION_UTILITIES)));
WK_EXCLUDED_APPLICATION_UTILITIES_FILES_ = WKApplicationUtilities.swift;
```

(also, added semicolons)

bnham replied:
Will do, seems like the proper variable here would be `WK_EXCLUDED_APPLICATION_UTILITIES_FILES_NO`

bnham committed:
Call into WKApplicationUtilities when handling certain types of links
https://bugs.webkit.org/show_bug.cgi?id=265172
rdar://118667681

Reviewed by Brent Fulgham and Elliott Williams.

In order to implement a particular feature, we need to call into WKApplicationUtilities when
handling certain types of navigation actions. This then calls a Swift-only API. To do this, we copy
WKApplicationUtilities.swift from WebKitAddition's include location to WebKit's DerivedSources. We
use EXCLUDED_SOURCE_FILES to exclude that file on platforms and SDKs where it doesn't exist.

This also requires us to know the top frame's origin at the point of navigation, and I couldn't find
any good way of accessing that state in decidePolicyForNavigationAction. So I modified
NavigationActionData to carry the topOrigin state.

* Source/WebKit/Configurations/WebKitSwift.xcconfig:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/NavigationActionData.h:
* Source/WebKit/Shared/NavigationActionData.serialization.in:
* Source/WebKit/UIProcess/Cocoa/NavigationState.mm:
(WebKit::NavigationState::NavigationClient::decidePolicyForNavigationAction):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::createWindow):
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
(WebKit::WebFrameLoaderClient::dispatchDecidePolicyForNavigationAction):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::didSameDocumentNavigationForFrameViaJSHistoryAPI):
(WebKit::WebLocalFrameLoaderClient::dispatchDecidePolicyForNewWindowAction):

Canonical link: https://commits.webkit.org/271405@main
